File: app/src/main/java/opencrowd/hgc/hgcwallet/hapi/tasks/UpdateTransactionsTask.java
package opencrowd.hgc.hgcwallet.hapi.tasks;

import com.hederahashgraph.api.proto.java.Query;
import com.hederahashgraph.api.proto.java.Response;
import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
import com.hederahashgraph.api.proto.java.TransactionRecord;

import java.util.List;

import opencrowd.hgc.hgcwallet.hapi.APIBaseTask;
import opencrowd.hgc.hgcwallet.hapi.APIRequestBuilder;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.database.DBHelper;
import opencrowd.hgc.hgcwallet.database.account.Account;
import opencrowd.hgc.hgcwallet.modals.HGCAccountID;

public class UpdateTransactionsTask extends APIBaseTask {
    public UpdateTransactionsTask() {
        super();
    }

    @Override
    public void main() {
        super.main();
        List<Account> accounts = DBHelper.getAllAccounts();
        if (accounts != null && !accounts.isEmpty()) {
            Account payer = accounts.get(0);
            if (payer.accountID() == null) {
                return;
            }
            for (Account account : accounts) {
                if (account.accountID() != null) {

                    try {
                        long fee = getCostOfGetAccountRecords(payer, account.accountID());
                        Query query = APIRequestBuilder.requestForGetAccountRecord(payer, account.accountID(),fee, node.accountID());
                        log(query.toString());
                        Response response = cryptoStub().getAccountRecords(query);
                        log(response.toString());
                        ResponseCodeEnum precheck = response.getCryptogetAccountBalance().getHeader().getNodeTransactionPrecheckCode();
                        switch (precheck) {
                            case OK:
                                List<TransactionRecord> records = response.getCryptoGetAccountRecords().getRecordsList();
                                if (records != null) {
                                    for (TransactionRecord record : records) {
                                        DBHelper.createTransaction(record, account.accountID());
                                    }
                                }
                                break;
                            case INVALID_ACCOUNT_ID:
                                error = "Invalid account " + account.accountID().stringRepresentation();
                                break;

                            default:
                                error = Singleton.INSTANCE.getErrorMessage(precheck);
                        }

                    } catch (Exception e) {
                        e.printStackTrace();
                        if (error != null)
                            error = getMessage(e);
                    }
                }
            }
        }
    }

    private long getCostOfGetAccountRecords(Account payer, HGCAccountID accountID) throws Exception {
        Query query = APIRequestBuilder.requestForGetAccountRecordCost(payer, accountID, node.accountID());
        log(query.toString());
        Response response = cryptoStub().getAccountRecords(query);
        log(response.toString());
        ResponseCodeEnum precheck = response.getCryptogetAccountBalance().getHeader().getNodeTransactionPrecheckCode();
        switch (precheck) {
            case OK:
                return response.getCryptoGetAccountRecords().getHeader().getCost();

            case INVALID_ACCOUNT_ID:
                error = "Invalid account " + accountID.stringRepresentation();
                throw new Exception();

            default:
                error = Singleton.INSTANCE.getErrorMessage(precheck);
                throw new Exception();
        }
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/hapi/tasks/UpdateTransactionsTask.kt
package hedera.hgc.hgcwallet.hapi.tasks

import com.hederahashgraph.api.proto.java.ResponseCodeEnum
import hedera.hgc.hgcwallet.Config
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.database.DBHelper
import hedera.hgc.hgcwallet.database.account.Account
import hedera.hgc.hgcwallet.hapi.APIBaseTask
import hedera.hgc.hgcwallet.hapi.GetAccountRecordParam
import hedera.hgc.hgcwallet.modals.HGCAccountID

class UpdateTransactionsTask : APIBaseTask() {

    override fun main() {
        super.main()
        getHistory(DBHelper.getAllAccounts())
    }


    private fun getHistory(accounts: List<Account>) {
        if (accounts.isNotEmpty()) {
            val payer = accounts.first()
            if (payer.accountID() == null) {
                return
            }
            for (account in accounts) {
                account.accountID()?.let { accID ->
                    try {
                        val fee = getCostOfAccountRecords(payer, accID)
                        val recordParam = GetAccountRecordParam(accID, fee)
                        val txnBuilder = payer.getTransactionBuilder()
                        val (query, response) = grpc.perform(recordParam, txnBuilder)
                        val precheck = response.cryptoGetAccountRecords.header.nodeTransactionPrecheckCode
                        when (precheck) {
                            ResponseCodeEnum.OK -> {
                                response.cryptoGetAccountRecords.recordsList?.let { records ->
                                    records.forEach {
                                        DBHelper.createTransaction(it)
                                        Singleton.setExchangeRate(it.receipt)
                                    }
                                }
                            }
                            ResponseCodeEnum.PAYER_ACCOUNT_NOT_FOUND -> {
                                error = "payerAccountNotFound " + payer.accountID()?.stringRepresentation()
                                throw Exception()
                            }

                            ResponseCodeEnum.INVALID_ACCOUNT_ID -> {
                                error = "invalidAccountID " + accID.stringRepresentation()
                                throw Exception()
                            }

                            else -> {
                                error = Singleton.getErrorMessage(precheck)
                                throw Exception()
                            }
                        }

                    } catch (e: Exception) {
                        if (error == null)
                            error = getMessage(e)

                        e.printStackTrace()
                    }

                }
            }
        }
    }

    @Throws(Exception::class)
    private fun getCostOfAccountRecords(payer: Account, accountID: HGCAccountID): Long {
        val accountRecordParam = GetAccountRecordParam(accountID)
        val txnBuilder = payer.getTransactionBuilder()
        val (query, response) = grpc.perform(accountRecordParam, txnBuilder)

        val status = response.cryptoGetAccountRecords.header.nodeTransactionPrecheckCode
        when (status) {
            ResponseCodeEnum.OK -> return response.cryptoGetAccountRecords.header.cost

            ResponseCodeEnum.PAYER_ACCOUNT_NOT_FOUND -> {
                error = "payerAccountNotFound " + payer.accountID()?.stringRepresentation()
                throw Exception(error)
            }

            ResponseCodeEnum.INVALID_ACCOUNT_ID -> {
                error = "invalidAccountID " + accountID.stringRepresentation()
                throw Exception(error)
            }

            else -> {
                error = Singleton.getErrorMessage(status)
                throw Exception(error)
            }
        }
    }
}
